### PR TITLE
Add additional `Input/Output` checks 

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

This PR adds additional input/output checks to transaction verification in order to cover all cases. Previously we added length checks in https://github.com/AleoNet/snarkVM/pull/2511; for completeness, we now add variant type checks as well.

CI is run [here](https://app.circleci.com/pipelines/github/ProvableHQ/snarkVM/413/workflows/c0c98765-3e0d-4b56-8be6-b255bdd6d8ec).

## Test Plan

Tested on local devnets and isonets. Sync tests have also been run to ensure that existing networks are fully compatible.

### CODE REVIEW

**Review by vicsn — APPROVED** [review]

**Review by d0cd — APPROVED** [review]

File: console/program/src/data_types/value_type/mod.rs [hunk 1]
```diff
@@ -38,6 +40,20 @@ pub enum ValueType<N: Network> {
     Future(Locator<N>),
 }
 
+impl<N: Network> ValueType<N> {
+    /// Returns the variant of the value type.
+    pub const fn variant(&self) -> Variant {
```
Comment (ljedrz) [inline console/program/src/data_types/value_type/mod.rs:45]: nit: the same could be achieved with `EnumIndex`, which is already derived for `ValueType`

File: synthesizer/process/src/verify_execution.rs [hunk 2]
```diff
@@ -112,6 +112,14 @@ impl<N: Network> Process<N> {
             ensure!(function.inputs().len() == num_inputs, "The number of transition inputs is incorrect");
             ensure!(function.outputs().len() == num_outputs, "The number of transition outputs is incorrect");
 
+            // Ensure the input and output types are equivalent to the ones defined in the function.
+            // We only need to check that the variant type matches because we already check the hashes in
+            // the `Input::verify` and `Output::verify` functions.
+            let transition_input_variants = transition.inputs().iter().map(Input::variant).collect::<Vec<_>>();
```
Comment (ljedrz) [inline synthesizer/process/src/verify_execution.rs:118]: nit: these allocations could be avoided via something like
```
ensure!(function.input_variants().iter().zip(transition.inputs().iter().map(Input::variant)).all(|(v1, v2)| v1 == v2));
```
(length equality has already been checked at this point)

File: synthesizer/program/src/function/mod.rs [hunk 3]
```diff
@@ -69,6 +69,11 @@ impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> Fun
         self.inputs.iter().map(|input| input.value_type()).cloned().collect()
     }
 
+    /// Returns the function input type variants.
+    pub fn input_variants(&self) -> Vec<Variant> {
```
Comment (ljedrz) [inline synthesizer/program/src/function/mod.rs:73]: nit: these could return an `impl Iterator<Item = Variant>`

**Review by ljedrz — APPROVED** [review]

**Review by zkxuerb — APPROVED** [review]
LGTM

File: synthesizer/program/src/function/mod.rs (hunk 3 above)
Comment (raychu86) [reply synthesizer/program/src/function/mod.rs:73]: I left it consistent with the existing `output_types` and `input_types` functions. We can revisit in the future for possible optimizations.

File: console/program/src/data_types/value_type/mod.rs (hunk 1 above)
Comment (raychu86) [reply console/program/src/data_types/value_type/mod.rs:45]: I added a `Variant` for consistency and to be more explicit in matching the `Variant` between types. I felt it was a bit vague to use EnumIndex.

File: synthesizer/process/src/verify_execution.rs (hunk 2 above)
Comment (raychu86) [reply synthesizer/process/src/verify_execution.rs:118]: ditto below: can leave for future optimization

**Pull request merged**